apps/zotonic_mod_menu/src/mod_menu.erl

```diff
@@ -226,88 +265,82 @@ create_new(List, Context) ->
-        {RscId, Context1}.
-
-
-
-%% @doc Notifier handler to get all menu ids for the default menu.
-observe_menu_get_rsc_ids(menu_get_rsc_ids, Context) ->
-    Menu = get_menu(Context),
-    {ok, menu_ids(Menu, [])}.
-
-    menu_ids([], Acc) ->
-        Acc;
-    menu_ids([{Id,SubMenu}|T], Acc) ->
-        Acc1 = menu_ids(SubMenu, [Id|Acc]),
-        menu_ids(T, Acc1);
-    menu_ids([H|T], Acc) ->
-        menu_ids(T, [H|Acc]).
+create_new([], Acc, Context) ->
+    {lists:reverse(Acc), Context};
+create_new([{Id,Sub}|Rest], Acc, Context) ->
+    {Id1, Context1} = create_new_if_needed(Id, Context),
+    {Sub1, Context2} = create_new(Sub, Context1),
+    create_new(Rest, [{Id1,Sub1}|Acc], Context2).
 
 
-%% @doc Observer the 'menu_save' notification
-observe_menu_save(#menu_save{id=MenuId, tree=Menu}, Context) ->
-    set_menu(MenuId, Menu, Context).
-
+create_new_if_needed(Id, Context) when is_integer(Id) ->
+    {Id, Context};
+create_new_if_needed(Id, Context) ->
+    C = lists:last( string:tokens(Id, "-") ),
+    Category = case m_category:name_to_id(C, Context) of
+        {error, _} -> text;
+        {ok, C1} -> C1
+    end,
+    Props = #{
+        <<"is_published">> => false,
+        <<"category">> => Category,
+        <<"title">> => ?__("New page", Context)
+    },
+    {ok, RscId} = m_rsc:insert(Props, Context),
+    OuterId = iolist_to_binary(["outernew-",integer_to_list(RscId)]),
+    InnerId = iolist_to_binary(["innernew-",integer_to_list(RscId)]),
+    Context1 = z_render:wire(
+        [
+            {script, [{script, ["$('#",Id,"').attr('id','",OuterId,"');"]}]},
+            {dialog_edit_basics, [
+                        {id, RscId},
+                        {target, InnerId},
+                        {template, "_menu_edit_item.tpl"}
+                ]},
+            {update, [
+                        {target, OuterId},
+                        {menu_id, InnerId},
+                        {id, RscId},
+                        {template, "_menu_edit_item.tpl"}
+                ]}
+        ], Context),
+    {RscId, Context1}.
+
+menu_ids([], Acc) ->
+    Acc;
+menu_ids([ #rsc_tree{ id = Id, tree = SubMenu } | T ], Acc) ->
+    Acc1 = menu_ids(SubMenu, [Id|Acc]),
+    menu_ids(T, Acc1).
 
 %% @doc Fetch the default menu. Performs validation/visibility checking on the menu items.
-%% @spec get_menu(Context) -> list()
 get_menu(Context) ->
     get_menu(m_rsc:rid(main_menu, Context), Context).
 
 %% @doc Fetch a menu structure from a rsc. Performs validation/visibility checking on the menu items.
-%% @spec get_menu(Id, Context) -> list()
 get_menu(Id, Context) ->
     case m_rsc:p(Id, menu, Context) of
         undefined -> [];
         <<>> -> [];
         Menu -> remove_invisible(validate(Menu, []), [], Context)
     end.
 
-	validate([], Acc) ->
-		lists:reverse(Acc);
-	validate([{_M,_L} = M|Ms], Acc) ->
-		validate(Ms, [M|Acc]);
-	validate([M|Ms], Acc) ->
-		validate(Ms, [{M,[]}|Acc]).
+validate(undefined, Acc) ->
+    Acc;
```

Comment:
Shouldn't this be a reverse too? Like the line below.

Reply:
Yes, probably better.